### PR TITLE
Add `Aabb` calculation for `Sprite`, `TextureAtlasSprite` and `Mesh2d`

### PR DESCRIPTION
**Pull request by opstic**
# Objective

- Add `Aabb` calculation for `Sprite`, `TextureAtlasSprite` and `Mesh2d`.
- Enable frustum culling for 2D entities since frustum culling requires a `Aabb` component in the entity to function.
- Improve 2D performance massively when there are many sprites out of view. (ex: `many_sprites`)

## Solution

- Derived from @Weasy666's #3944 pull request, which had no activity since multiple months.
- Adapted the code to the latest version of Bevy.
- Added support for sprites with non-center `Anchor`s to avoid culling prematurely when part of the sprite is still in view or not culling when sprite is already out of view.

### Note
- Gives 15.8x performance boosts in some scenarios. (5 fps vs 79 fps with 409600 sprites in `many_sprites`)

### CODE REVIEW

**Comment by james7132** [comment]
Wondering if we should have a `Aabb2D` to reduce the memory usage and computation cost of running this. Not blocking, but probably a good idea as an optimization.

**Comment by Elabajaba** [comment]
This gave me a huge fps boost on `many_sprites` with the default settings (102400 sprites), going from ~60fps to ~1100fps.

Bevymark performance seemed slightly faster with 10 batches of 10000 sprites (100k total), though that could've just been the normal run to run variation since I didn't bother closing everything first to reduce noise.

**Review by Elabajaba — APPROVED** [review]

File: crates/bevy_sprite/src/lib.rs [hunk 1]
```diff
@@ -86,3 +94,51 @@ impl Plugin for SpritePlugin {
...
+    for (entity, mesh_handle) in meshes_without_aabb.iter() {
+        if let Some(mesh) = meshes.get(&mesh_handle.0) {
+            if let Some(aabb) = mesh.compute_aabb() {
+                commands.entity(entity).insert(aabb);
+            }
+        }
+    }
+    for (entity, sprite, texture_handle) in sprites_without_aabb.iter() {
+        if let Some(image) = images.get(texture_handle) {
+            let size = sprite.custom_size.unwrap_or_else(|| image.size());
```
Comment (devil-ira) [inline crates/bevy_sprite/src/lib.rs:122]:
```suggestion
        if let Some(size) = sprite
            .custom_size
            .or_else(|| images.get(texture_handle).map(|image| image.size()))
        {
```
Avoids the asset lookup when `custom_size` is `Some`

File: crates/bevy_sprite/src/lib.rs [hunk 2]
```diff
@@ -86,3 +94,51 @@ impl Plugin for SpritePlugin {
...
+        if let Some(atlas) = atlases.get(atlas_handle) {
+            if let Some(rect) = atlas.textures.get(atlas_sprite.index) {
+                let size = atlas_sprite
+                    .custom_size
+                    .unwrap_or_else(|| (rect.min - rect.max).abs());
+                let aabb = Aabb {
+                    center: (-atlas_sprite.anchor.as_vec() * size).extend(0.0).into(),
+                    half_extents: (0.5 * size).extend(0.0).into(),
+                };
+                commands.entity(entity).insert(aabb);
```
Comment (devil-ira) [inline crates/bevy_sprite/src/lib.rs:140]:
```suggestion
        if let Some(size) = atlas_sprite.custom_size.or_else(|| {
            atlases
                .get(atlas_handle)
                .and_then(|atlas| atlas.textures.get(atlas_sprite.index))
                .map(|rect| (rect.min - rect.max).abs())
        }) {
            let aabb = Aabb {
                center: (-atlas_sprite.anchor.as_vec() * size).extend(0.0).into(),
                half_extents: (0.5 * size).extend(0.0).into(),
            };
            commands.entity(entity).insert(aabb);
```
Avoids the asset lookup when `custom_size` is `Some`

**Review by devil-ira — COMMENTED** [review]
I'm surprised this doesn't negatively impact performance when all sprites are on screen like in bevymark, but that's good :)

**Comment by Weasy666** [comment]
Oh... i absolutely forgot that my PR stil existed, if would just have pinged me, i would have invested the time to bring it up to date 🙃.
But I have absolutely nothing against the fact that someone else has invested the time to get this done 😆. I would just be happy to also get some kind of attribution 🙂🙈

**Comment by opstic** [comment]
> I would just be happy to also get some form of attribution 🙂🙈

I have this line in the main section of this pull request:
"Derived from @Weasy666's 2d frustum culling #3944 pull request, which had no activity since multiple months."
Shouldn't this be included in the commit if this was merged? Because I've seen this behavior for other pull requests.
If that's not enough, please tell me how to add more attribution. I'm pretty new to contributing, after all 😅.

File: crates/bevy_sprite/src/lib.rs (hunk 2 above)
Comment (opstic) [reply crates/bevy_sprite/src/lib.rs]: Thank you!

**Review by devil-ira — APPROVED** [review]

**Pull request merged**